Reapply "Load Microsoft.DotNet.MSBuildSdkResolver into default load context"

ladipro opened the pull request:
### Context

Now that https://github.com/dotnet/sdk/pull/39573 has made it into VS, we can finally enable loading `Microsoft.DotNet.MSBuildSdkResolver` by assembly name, as opposed to file path, which makes it possible to use its NGEN image.

### Changes Made

Re-did #9439 which had to be reverted because of the problematic dependency on `Newtonsoft.Json`. The .NET SDK resolver does not depend on `Newtonsoft.Json` anymore. All the other pieces are already in place:
- `Microsoft.DotNet.MSBuildSdkResolver` is NGENed, both for MSBuild.exe and for devenv.exe.
- `devenv.exe.config` contains binding redirects analogous to what's added in this PR.

### Testing

Experimentally inserted into VS and verified that `Microsoft.DotNet.MSBuildSdkResolver` is no longer JITted, saving ~50-100 ms of MSBuild.exe startup time.

JanKrivanek commented:
Can you please link the mentioned exp VS insertion?

ladipro commented:
> Can you please link the mentioned exp VS insertion?

https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/575724 though it's odd that it hasn't run Perf DDRITs. I just requested Perf DDRITs one more time. If it doesn't work I'll create a new insertion.

JanKrivanek commented on src/MSBuild/app.amd64.config:
```diff
@@ -184,6 +184,17 @@
           <codeBase version="17.0.0.0" href="..\..\..\Microsoft\VisualStudio\v17.0\CodeAnalysis\Microsoft.VisualStudio.CodeAnalysis.Sdk.dll" />
         </dependentAssembly>
+
+        <!-- Redirects for SDK resolver components -->
```
Can the https://github.com/dotnet/msbuild/blob/main/documentation/NETFramework-NGEN.md#microsoftdotnetmsbuildsdkresolver be linked here?
I was confused why the version is fixed here - so the doc can help clarify it.

ladipro replied:
I was a bit hesitant about including a GitHub link in a file that gets placed on end user machines unscrubbed. Especially since this is not a perma-link. But it's potentially useful for posterity. Done, thank you!

JanKrivanek commented on src/MSBuild/app.config:
```diff
@@ -137,6 +137,17 @@
           <codeBase version="17.0.0.0" href="..\..\Microsoft\VisualStudio\v17.0\CodeAnalysis\Microsoft.VisualStudio.CodeAnalysis.Sdk.dll" />
         </dependentAssembly>
+
+        <!-- Redirects for SDK resolver components -->
```
Same here - the link to doc (or explanative comment) can save future time

JanKrivanek reviewed (approved):
Looks good.

I have minore comment for improving maintainability

ladipro commented:
I have addressed the comment and made another exp insertion, this time with Perf DDRITs results. It seems to be clean module some minor unrelated noise.
https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/575802

rainersigwald reviewed (approved):
This looks good to me but I was surprised to not see an improvement in the MethodsJitted counters in the insertion. Are we sure it's working?

ladipro commented:
> This looks good to me but I was surprised to not see an improvement in the MethodsJitted counters in the insertion. Are we sure it's working?

I have high confidence that JITting of `Microsoft.DotNet.MSBuildSdkResolver` is eliminated in MSBuild.exe invoked from the command line. Now for PerfDDRITs, the CLR_AdjustedMethodsJitted_Total_NonDevenv counter is not showing any JITting in the resolver, neither baseline nor target build. This may be because the scenarios that measure this counter don't really cover startup / first evaluation. Or there's something else at play. My memory is faint, but isn't SDK resolution (in VS only?) optimized by forwarding requests from worker nodes back to the central node?
